pandas/tests/series/indexing/test_setitem.py

```diff
@@ -158,7 +158,7 @@ def test_setitem_series_object_dtype(self, indexer, ser_index):
         expected = Series([Series([42], index=[ser_index]), 0], dtype="object")
         tm.assert_series_equal(ser, expected)
 
-    @pytest.mark.parametrize("index, exp_value", [(0, 42.0), (1, np.nan)])
+    @pytest.mark.parametrize("index, exp_value", [(0, 42), (1, np.nan)])
     def test_setitem_series(self, index, exp_value):
```

Comment:
this seems weird.  wouldnt we expect an object-dtype Series with ser.loc[0] itself being a Series?

Reply:
that's unrelated to the changes here. so will not be changing the indexing code in this PR. This PR is a followon to #40555 to change can_hold_element, which I'm uncomfortable doing on 1.2.x since we may not have another release after the next one.

Reply:
OK.  cc @phofl thoughts on desired behavior?

Reply:
I think the behavior after this pr is correct.
```
ser.loc[0] = Series([42], index=[0])
```

This assigns 42 to the first element in the Series, so would not expect a type cast here

Reply:
so you expect to get the same result from `ser.loc[0] = ser2` as you would from `ser.loc[[0]] = ser2`?

Reply:
Hm good point. Not sure how alignment should work there exactly. Are there any cases where this would assign a Series as first element?

Reply:
did this resolve? (is this a separate patch / issue s well)?

Reply:
we need a separate issue/patch if any other changes. @jbrockmendel @phofl 

if another patch is done, this PR is not urgent as it becomes just a refactor.

if another patch is not done, we should merge this before 1.3.

Reply:
@jbrockmendel @phofl is there a plan to change the return type. otherwise we should probably merge this for 1.3

waiting and we would need a release note and decide on backport (which i'd be -1 on, for the same reason I didn't change can_hold_element in 1.2.5)

Reply:
Lgtm too. Even if we agree that we should return something else, the int result is more consistent than the float result

Reply:
The correct-behavior question may be related to https://github.com/pandas-dev/pandas/issues/16864